lib/services/composer.js

```diff
@@ -5,58 +5,108 @@ const _ = require('lodash'),
   components = require('./components'),
   references = require('./references'),
   mapLayoutToPageData = require('../utils/layout-to-page-data'),
-  referenceProperty = '_ref';
+  { getSchema } = require('../utils/schema'),
+  { isLayout, isComponent } = require('clayutils'),
+  referenceProperty = '_ref',
+  renderOrderProperty = 'renderOrder',
+  defaultRenderOrder = 1;
 var log = require('./logger').setup({ file: __filename });
 
 /**
  * Compose a component, recursively filling in all component references with
  * instance data.
+ * @param {string} uri
  * @param {object} data
  * @param {object} locals  Extra data that some GETs use
  * @param {function|string} [filter='_ref']
  * @returns {Promise} - Resolves with composed component data
  */
-function resolveComponentReferences(data, locals, filter = referenceProperty) {
-  const referenceObjects = references.listDeepObjects(data, filter);
-
-  return bluebird.all(referenceObjects).each(referenceObject => {
-    return components.get(referenceObject[referenceProperty], locals)
-      .then(obj => {
-        // the thing we got back might have its own references
-        return resolveComponentReferences(obj, locals, filter).finally(() => {
-          _.assign(referenceObject, _.omit(obj, referenceProperty));
-        }).catch(function (error) {
-          const logObj = {
-            stack: error.stack,
-            cmpt: referenceObject[referenceProperty]
-          };
-
-          if (error.status) {
-            logObj.status = error.status;
-          }
-
-          log('error', `${error.message}`, logObj);
-
-          return bluebird.reject(error);
+function resolveComponentReferences(uri, data, locals, filter = referenceProperty) {
```

Comment:
adding `uri` to the beginning of this is definitely a breaking change if anyone is using the composer externally. i added it to the front here because that seems to be the norm across the repo. but i can move it to the end and make it optional to clear the breaking change if that's preferred